pkg/templating/k8s_validation.go

```diff
@@ -0,0 +1,388 @@
+func validateString(s string, validator func(string, int) bool) []int {
+	var failedIndexes []int
+
+	length := len(s)
+
+	for idx := 0; idx < length; idx++ {
+		if !validator(s, idx) {
+			failedIndexes = append(failedIndexes, idx)
+		}
+	}
+
+	return failedIndexes
+}
+
+// Validates and copies the string (possibly modified) to a new byte array
+//
+// Needs a validator function and a copier function.
+// The validator function is passed straight to validateString
+// The copier function receives the byte array being built, the string, its index,
+// and whether the source character passed validation, and returns the (potentially
+// reallocated) byte array and whether the copied character was valid.
+//
+// Returns the copied and possibly modified string, and the list of source positions where
+// validation errors were found.
+func validateAndCopyString(s string, validator validationCheckerFn, copier validationCopierFn) (string, []int) {
+	newValue := make([]byte, 0, len(s))
+
+	failedIndexes := validateString(s, func(s string, idx int) bool {
+		ok := validator(s, idx)
+		ok, newValue = copier(newValue, s, idx, ok)
+
+		return ok
+	})
+
+	return string(newValue), failedIndexes
+}
+
+// takes a user supplied length and ensures it is within bounds else return the forced vale
+func maxLenEnforcer(suppliedLen, minBound, forceTo int) int {
+	if suppliedLen < minBound || suppliedLen > forceTo {
+		return forceTo
+	}
+	return suppliedLen
+}
+
+// returns a validationCheckerFn closure required by k8sDNS1123LabelValidation
+func getK8DNS1123LabelValidationCheckerFN(length int) validationCheckerFn {
+	return func(s string, idx int) bool {
+		c := s[idx]
+
+		// check the edges, beginning and ending of the string
+		if idx == 0 || idx == length-1 {
+			return isLowerAlphaNum(c)
+		}
+
+		// check the body
+		return isLowerAlphaNumDash(c)
+	}
+}
+
+// returns a validationCopierFn closure required by k8sDNS1123LabelValidation
+func getK8DNS1123LabelValidationCopierFN(length int, fixup bool) validationCopierFn {
+	return func(byteSlice []byte, s string, idx int, ok bool) (bool, []byte) {
+		var c byte
+
+		if ok || !fixup {
+			c = s[idx]
+		} else {
+			// can always fix this
+			ok = true
+
+			if idx == 0 || idx == length-1 {
+				c = dns1123LabelValueEdgeFix
+			} else {
+				c = dns1123LabelValueBodyFix
+			}
+		}
+
+		return ok, append(byteSlice, c)
+	}
+}
+
+// returns a validationCheckerFn closure required by K8sLabelValueValidation
+func getK8LabelValidationCheckerFN(length int) validationCheckerFn {
+	return func(s string, idx int) bool {
+		c := s[idx]
+
+		// check the edges, beginning and ending of the string
+		if idx == 0 || idx == length-1 {
+			return isAlphaNum(c)
+		}
+
+		// check the body
+		return isAlphaNumDashDotUnderscore(c)
+	}
+}
+
+// returns a validationCopierFn closure required by K8sLabelValueValidation
+func getK8LabelValidationCopierFN(length int, fixup bool) validationCopierFn {
```

Comment:
maybe we can merge this and the other closure generation by just passing it the characters they need to replace?